File: src/blockchat_api/src/blockchain_module/block.rs
use crate::blockchain_module::transaction::Transaction;
use sha256::digest;
use serde::{Serialize, Deserialize};
use serde_json;

#[derive(Clone,Serialize, Deserialize, Debug)]
pub struct Block {
    pub index: i32,
    pub timestamp: i64,
    pub transactions: Vec<Transaction>,
    pub validator: i32,
    pub current_hash: String,
    pub previous_hash: i32,
}

impl Block { 
    
    // Constructor function for struct Block
    pub fn new(index: i32, timestamp: i64, transactions: Vec<Transaction>,
                validator: i32, current_hash: String, previous_hash: i32) -> Self {
        Block { 
            index,
            timestamp,
            transactions,
            validator,
            current_hash,
            previous_hash
        }
    }

    // Jsonify Block Object 
    pub fn jsonify(&self) -> Result<String, serde_json::Error> {
        serde_json::to_string(self)
    }

    // Create a hash for the Block Object
    pub fn hashify(&mut self) {
        let json = self.jsonify();
        match json {
            Ok(json) => self.current_hash = digest(json),
            Err(_e) => println!("Could not create hash"),
        }
    }
    
    // Adding a transaction in a block, not exceeding its capacity
    pub fn add_transaction(&mut self, transaction: Transaction, capacity: usize) {
        if self.transactions.len() < capacity {
            self.transactions.push(transaction);
        }
    }

}

File: src/API/blockchain/block.go
package blockchain

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"time"
)

// Block represents each 'item' in the blockchain
type Block struct {
    Index         int           `json:"index"`
    Timestamp     int64         `json:"timestamp"`
    Transactions  []Transaction `json:"transactions"`
    Validator     int           `json:"validator"`
    CurrentHash   string        `json:"current_hash"`
    PreviousHash  string        `json:"previous_hash"`
}

// NewBlock creates and returns a new Block
func NewBlock(index int, previousHash string) *Block {
    t := time.Now();
    timestamp := t.UnixNano();
    return &Block{
        Index:         index,
        Timestamp:     timestamp,
        Transactions:  nil,
        Validator:     0,
        PreviousHash:  previousHash,
        CurrentHash:   "",
    }
}

// JSONify serializes the Block into a JSON string
func (b *Block) JSONify() (string, error) {
    jsonBytes, err := json.Marshal(b)
    return string(jsonBytes), err
}

// Hashify updates the current hash of the block
func (b *Block) Hashify() {
    jsonString, err := b.JSONify()
    if err != nil {
        fmt.Println("Could not create hash:", err)
        return
    }
    hash := sha256.Sum256([]byte(jsonString))
    b.CurrentHash = hex.EncodeToString(hash[:])
}

// AddTransaction adds a new transaction to the block if there's capacity
func (b *Block) AddTransaction(transaction *Transaction, capacity int) {
    if len(b.Transactions) < capacity {
        b.Transactions = append(b.Transactions, *transaction)
    }
}

// Set previous Hash of a block
func (b *Block) GetPreviousHash(chain *Blockchain) {
    b.PreviousHash = chain.Chain[len(chain.Chain) - 1].CurrentHash;
}

